### ISSUE TITLE
Validation can be called on a node that was deleted by previous validations

### ISSUE THREAD

**Issue by Soreine** [ianstormtaylor/slate#1378]
I'm currently using the new schema to write validations. But I am encountering an obscure bug. During validation, there is probably a call to `validateNode`  on an outdated node. In the case described below, an error is thrown during normalization, because `validateNode` is called on a node that was deleted by previous calls.

Here is a minimal working file that reproduces the issue.

```jsx
/** @jsx h */
import { createHyperscript } from 'slate-hyperscript';
import { Schema } from 'slate';

const h = createHyperscript({
    blocks: {
        paragraph: 'paragraph',
        quote: 'quote',
        list: 'list',
        item: 'item'
    }
});

const plugin = {
    schema: {
        blocks: {
            // Quotes only contain paragraphs
            quote: {
                nodes: [{ types: ['paragraph'] }]
            },

            // Items should be inside a list. Wrap them if needed.
            list: {},
            item: {
                parent: { types: ['list'] },
                normalize(change, reason, context) {
                    if (reason === 'parent_type_invalid') {
                        // Wrap item in a list
                        change.wrapBlockByKey(context.node.key, 'list');
                    }
                }
            }
        }
    }
};

const schema = Schema.create({
    plugins: [plugin]
});

// Value to normalize:

let value = (
    <value>
        <document>
            <quote>
                <item>
                    This item should be wrapped in a list. But a quote can only
                    contain paragraph, in the end it should be deleted, as well
                    as the whole quote.
                </item>
            </quote>
            <quote>
                <paragraph>Valid paragraph</paragraph>
                <item key="not-found">
                    Again, this should be deleted. But the quote should remain
                </item>
            </quote>
        </document>
    </value>
);

// Attach the schema to the test value
value = value.set('schema', schema);
// Normalize the value
const normalized = value.change().normalize().value; 
// ERROR: Could not find a descendant node with key "not-found".
```

For information, this is the expected value out of the normalization:
```jsx
// Expected value out of normalization:
const expectedNormalized = (
    <value>
        <document>
            <quote>
                <paragraph>Valid paragraph</paragraph>
            </quote>
        </document>
    </value>
);
```

I noticed that removing the first quote eliminates the error. The following value is properly normalized:
```jsx
// No error normlalizing this one
const noError = (
    <value>
        <document>
            <quote>
                <paragraph>Valid paragraph</paragraph>
                <item key="not-found">
                    This is properly deleted. And the quote remains.
                </item>
            </quote>
        </document>
    </value>
);
```

So maybe something like this is happening?

- Calls validate on `document`
  - Calls validate on first child of `document`: the first quote. 
    - Removes first quote => Calls validation on ancestors/siblings, which normalizes the second quote, and removes the item `not-found`.
  - End of first child of `document` validation
  - Calls validate on second child of `document`: the second quote. But here the value of the second quote node is outdated.

**Comment by Soreine** [comment]
Duplicate of #1363
